tsdb/engine/tsm1/engine.go

```diff
@@ -154,80 +154,87 @@ func (e *Engine) SetEnabled(enabled bool) {
 func (e *Engine) SetCompactionsEnabled(enabled bool) {
 	if enabled {
 		e.enableSnapshotCompactions()
-		e.enableLevelCompactions()
-
+		e.enableLevelCompactions(0)
 	} else {
 		e.disableSnapshotCompactions()
-		e.disableLevelCompactions()
+		e.disableLevelCompactions(0)
 	}
 }
 
-func (e *Engine) enableLevelCompactions() {
+func (e *Engine) enableLevelCompactions(n int) {
 	e.mu.Lock()
-	if e.levelCompactionsEnabled {
+	e.levelWorkers -= n
+	if e.levelWorkers != 0 || e.done != nil {
+		// still waiting on more workers or already enabled
 		e.mu.Unlock()
 		return
 	}
-	e.levelCompactionsEnabled = true
+
+	// last one to enable, start things back up
 	e.Compactor.EnabledCompactions()
-	e.done = make(chan struct{})
-	e.mu.Unlock()
+	quit := make(chan struct{})
+	e.done = quit
 
 	e.wg.Add(4)
-	go e.compactTSMFull()
-	go e.compactTSMLevel(true, 1)
-	go e.compactTSMLevel(true, 2)
-	go e.compactTSMLevel(false, 3)
+	e.mu.Unlock()
+
+	go func() { defer e.wg.Done(); e.compactTSMFull(quit) }()
+	go func() { defer e.wg.Done(); e.compactTSMLevel(true, 1, quit) }()
+	go func() { defer e.wg.Done(); e.compactTSMLevel(true, 2, quit) }()
+	go func() { defer e.wg.Done(); e.compactTSMLevel(false, 3, quit) }()
 }
 
-func (e *Engine) disableLevelCompactions() {
+func (e *Engine) disableLevelCompactions(n int) {
```

Comment:
`n` represents the number of workers that are expecting compactions to remain disabled. Since this can be called externally from `SetCompactionsEnabled`, and we can't guarantee the same call pattern as we can when used internally, I went with not tracking the status of external calls to enable/disable.

Thinking about it some more, we could try to track the state that the external calls think the status should be, and use that as a "worker"